Doc/tools/extensions/grammar_snippet.py

```diff
@@ -0,0 +1,192 @@
+    """Create a literal block from options & content.
+
+    This implements the common functionality for GrammarSnippetDirective
+    and CompatProductionList.
+    """
+
+    group_name = options['group']
+
+    # Docutils elements have a `rawsource` attribute that is supposed to be
+    # set to the original ReST source.
+    # Sphinx does the following with it:
+    # - if it's empty, set it to `self.astext()`
+    # - if it matches `self.astext()` when generating the output,
+    #   apply syntax highlighting (which is based on the plain-text content
+    #   and thus discards internal formatting, like references).
+    # To get around this, we set it to this non-empty string:
+    rawsource = 'You should not see this.'
+
+    literal = nodes.literal_block(
+        rawsource,
+        '',
+        # TODO: Use a dedicated CSS class here and for strings.
+        # and add it to the theme too
+        classes=['highlight'],
+    )
+
+    grammar_re = re.compile(
+        """
+            (?P<rule_name>^[a-zA-Z0-9_]+)     # identifier at start of line
+            (?=:)                             # ... followed by a colon
+        |
+            [`](?P<rule_ref>[a-zA-Z0-9_]+)[`] # identifier in backquotes
+        |
+            (?P<single_quoted>'[^']*')        # string in 'quotes'
+        |
+            (?P<double_quoted>"[^"]*")        # string in "quotes"
+        """,
+        re.VERBOSE,
+    )
+
+    for line in content:
+        last_pos = 0
+        for match in grammar_re.finditer(line):
+            # Handle text between matches
+            if match.start() > last_pos:
+                literal += nodes.Text(line[last_pos : match.start()])
+            last_pos = match.end()
+
+            # Handle matches
+            groupdict = {
+                name: content
+                for name, content in match.groupdict().items()
+                if content is not None
+            }
+            match groupdict:
+                case {'rule_name': name}:
+                    name_node = addnodes.literal_strong()
+
+                    # Cargo-culted magic to make `name_node` a link target
+                    # similar to Sphinx `production`.
+                    # This needs to be the same as what Sphinx does
+                    # to avoid breaking existing links.
+                    domain = directive.env.domains['std']
+                    obj_name = f"{group_name}:{name}"
+                    prefix = f'grammar-token-{group_name}'
+                    node_id = make_id(
+                        directive.env, directive.state.document, prefix, name
+                    )
+                    name_node['ids'].append(node_id)
+                    directive.state.document.note_implicit_target(
+                        name_node, name_node
+                    )
+                    domain.note_object(
+                        'token', obj_name, node_id, location=name_node
+                    )
```

Comment:
Could we factor this to a helper method? Is this code intended to be transitional or will it stay long-term?

Reply:
Sounds good!